src/bentoml/_internal/server/grpc/servicer/v1/__init__.py

```diff
@@ -72,20 +74,47 @@ async def Call(  # type: ignore (no async types) # pylint: disable=invalid-overr
                     validate_proto_fields(request.WhichOneof("content"), api.input),
                 )
                 input_data = await api.input.from_proto(input_proto)
+                ctx: InferenceApiContext | None = None
+                assert api.func is not None
+                # NOTE: function should always be set here, but check anyway.
                 if asyncio.iscoroutinefunction(api.func):
                     if api.multi_input:
+                        if api.needs_ctx:
+                            ctx = InferenceApiContext.from_grpc(context)
+                            input_data[api.ctx_param] = ctx
                         output = await api.func(**input_data)
                     else:
-                        output = await api.func(input_data)
+                        if api.needs_ctx:
+                            ctx = InferenceApiContext.from_grpc(context)
+                            output = await api.func(input_data, ctx)
+                        else:
+                            output = await api.func(input_data)
                 else:
                     if api.multi_input:
+                        if api.needs_ctx:
+                            ctx = InferenceApiContext.from_grpc(context)
+                            input_data[api.ctx_param] = ctx
                         output = await anyio.to_thread.run_sync(api.func, **input_data)
                     else:
-                        output = await anyio.to_thread.run_sync(api.func, input_data)
-
-                res = await api.output.to_proto(output)
+                        if api.needs_ctx:
+                            ctx = InferenceApiContext.from_grpc(context)
+                            output = await anyio.to_thread.run_sync(
+                                api.func, input_data, ctx
+                            )
+                        else:
+                            output = await anyio.to_thread.run_sync(
+                                api.func, input_data
+                            )
+
+                if ctx is not None:
+                    # Set gRPC metadata if context is available
+                    context.set_trailing_metadata(
+                        aio.Metadata.from_tuple(tuple(ctx.response.metadata.items()))
+                    )
```

Comment:
Hm, is this right? Wouldn't you be sending the trailing metadata twice? Or are users not meant to modify the gRPC context directly? (In which case we should expose our own metadata implementation)

Reply:
I only it send it once right? only if ctx is given. Usually, if user wants to handle context, either client/server interceptors should be used.

Context are meant to be modified if needed.

Reply:
But for the majority of users (exclude power user), context is not needed. It is the case only use if you know what you are doing.

Reply:
No, I mean with a "normal" gRPC server, does modifying the context do anything? We're both modifying the context itself and sending the metadata manually, we should do one or the other, I think.